Post model

mattegan111 opened the pull request.

mattegan111 committed:
Improve error message

Provide info to developer.

mattegan111 committed:
Add Post model

mattegan111 committed:
Add Post CRUD routes

nathanrobb commented on server/models/Post.js:
```diff
@@ -0,0 +1,28 @@
+	author: {
+		type: mongoose.Schema.Types.ObjectId,
+		required: true,
+	},
```
did this want a ref to user like in the likes? `ref: 'User',`?

nathanrobb commented on server/routes/posts.js:
```diff
@@ -1,32 +1,171 @@
+		const posts = await Post.find({
+			author: req.user.id,
+		}).sort({
+			date: -1,
```
What does `-1` mean in this context? Is it descending sort?

mattegan111 replied:
Yes :). I'll leave it as is, but I wonder, is something like this preferable for the sake of readability?
```suggestion
			date: 'desc',
```

nathanrobb replied:
Yeah, `'desc'` would help readability a ton.

nathanrobb commented on server/routes/posts.js:
```diff
@@ -1,32 +1,171 @@
+			// If 'like' is provided, it should not be empty
+			body(
+				'like',
+				'A User Id is required to like this post'
```
Is this comment not right?

mattegan111 replied:
As per our discussion:

This comment is as intended but there are wider issues that need addressing.

One issue is that the userId isn't a value entered by the user, it's stored on the req object, so it doesn't need to be validated anyway.

Another issue is that this route is trying to serve two purposes: 
1. Allow the author of a post to edit the post's bodytext
2. Allow any user to edit the post's like array, adding their userid to the array, hereby 'liking' the post.

These should just be separate routes. An alternative solution can result in endpoints with more logical nesting. I intend to do something like the following:

`app.post('/posts/:postId/likes')`
`app.delete('/posts/:postId/likes')`

My understanding is that this is a good endpoint structure regardless of the structure of our database.

nathanrobb commented on server/models/Post.js:
```diff
@@ -0,0 +1,28 @@
+		required: true,
+	},
+
+	author: {
```
Perhaps this could be named `authorId`?

mattegan111 replied:
Much better, will do.

mattegan111 replied:
Just out of interest, maybe one disadvantage of `authorId` is that you could have 
`
const post = Post.getById('id').populate('author');
`
`
const postAuthorName = post.authorId.name
`
Where it seems weird to have what's supposedly an 'id' but it's actually an object and you're looking at it's name.

nathanrobb commented on server/routes/posts.js:
```diff
@@ -1,32 +1,171 @@
-router.put('/', (req, res) => {
-	res.send('Update post');
-});
+router.put(
```
Did we want to support removing a like here or were we wanting to force that through the `DELETE` only?

mattegan111 replied:
Yes, let's support removing likes. In my comment above I mentioned making a route specifically for handling this:

`app.delete('/posts/:postId/likes')`

nathanrobb commented on server/routes/posts.js:
```diff
@@ -1,32 +1,171 @@
+		const errors = validationResult(req);
+		if (!errors.isEmpty()) {
+			return res
+				.status(400)
```
This is correct, but why are we using `400` here and a `404`/`401` status later on?

mattegan111 replied:
Here's what I learned:

All 4xx HTTP codes relate to an error on the user's part. 400 is a generic code whereas all other 4xx codes including 404 and 401 also have more specific meanings. 
Here we use 400 because the input from the user is unacceptable if it doesn't pass validation; we want to tell the user to try sending different information to us. There's no relevant 4xx code with more specificity so 400 is most appropriate. (422 is discussed as an option, but is a matter of contention and mostly not accepted).

404 basically means there's no match for the requested URL, appropriate when the user requests a post that doesn't exist - the URL which contains the intended post id leads nowhere. 

401 isn't used appropriately here. 401 'Unauthorised' is like the server saying to the user "Who are you? You should try authorising.".

403 'Forbidden' would be more appropriate. "I know who you are, but I'm not doing that."

nathanrobb reviewed (approved):
LGTM

mattegan111 committed:
Improve readability of descending by date

Co-authored-by: Nathan Robb <nathanrobb@hotmail.com>

mattegan111 committed:
Fix incorrect response status and message

These responses are more accurately described as 403 rather than 401.

mattegan111 committed:
Rename model key

Rename key because the db entry will contain an Id rather than the object itself.

mattegan111 committed:
Added missing ref

This way we can store a reference to another model so that at some point we can populate those to get the complete model by that reference.

mattegan111 committed:
Separate routes for adding/removing likes

We can adhere to a RESTful structure by having separate routes for adding a like to a post and deleting a like from a post - as opposed to an 'edit post' function that handles both of these as well as editing the text of a post.
